File: food-ordering-system/java-without-multithreading/strategy/LowestCostStrategy.java
package strategy;
import entities.Restaurant;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class LowestCostStrategy implements RestaurantSelectionStrategy {
    @Override
    public Optional<Restaurant> selectRestaurant(List<Restaurant> eligibleRestaurants, Map<String, Integer> requestedItems) {
        Restaurant bestRestaurant = null;
        double minCost = Double.MAX_VALUE;

        // Decoupled dynamic loop evaluating variable item amounts correctly mapped natively!
        for (Restaurant r : eligibleRestaurants) {
            double currentCost = 0;
            for (Map.Entry<String, Integer> entry : requestedItems.entrySet()) {
                currentCost += (r.getMenu().get(entry.getKey()) * entry.getValue());
            }

            if (currentCost < minCost) {
                minCost = currentCost;
                bestRestaurant = r;
            }
        }
        return Optional.ofNullable(bestRestaurant);
    }
}

File: food-ordering-system/go-without-multithreading/strategy/lowest_cost_strategy.go
package strategy
import (
	"app/food-ordering-system/go-without-multithreading/entities"
	"math"
)

type LowestCostStrategy struct{}

func (s *LowestCostStrategy) SelectRestaurant(eligibleRestaurants []*entities.Restaurant, requestedItems map[string]int) *entities.Restaurant {
	var bestRestaurant *entities.Restaurant
	minCost := math.MaxFloat64

	for _, r := range eligibleRestaurants {
		currentCost := 0.0
		for item, qty := range requestedItems {
			currentCost += (r.Menu[item] * float64(qty))
		}
		if currentCost < minCost {
			minCost = currentCost
			bestRestaurant = r
		}
	}
	return bestRestaurant
}
